File: ByLanguage/android/BookSample/HellloWorld/app/src/main/java/com/tq/myapplication/MainActivity.java
package com.tq.myapplication;

import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.util.Log;

public class MainActivity extends AppCompatActivity {

    private static final String DEBUG_TAG = "MyFristAppLogging";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        Log.i(DEBUG_TAG, "onCreate()");

        forceError();
    }

    public void forceError() {
        if (true) {
            throw new Error("Whoops");
        }
    }
}

File: ByLanguage/android/LifecycleDemo/app/src/main/java/com/tq/lifecycledemo/MainActivity.kt
package com.tq.lifecycledemo

import android.arch.lifecycle.Lifecycle
import android.arch.lifecycle.LifecycleObserver
import android.arch.lifecycle.OnLifecycleEvent
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.util.Log.d

class MyLifecycleObserver : LifecycleObserver {
    @OnLifecycleEvent(Lifecycle.Event.ON_RESUME)
    fun connectListener() {
        d("TEST", "RESUME")
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_PAUSE)
    fun disconnectListener() {
        d("TEST", "PAUSE")
    }
}

class MainActivity : AppCompatActivity() {

    var observer: MyLifecycleObserver = MyLifecycleObserver()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        lifecycle.addObserver(observer)
    }
}
